app/modules/node.c

```diff
@@ -171,69 +171,59 @@ static int node_heap( lua_State* L )
   return 1;
 }
 
-extern int  lua_put_line(const char *s, size_t l);
-extern bool user_process_input(bool force);
-
 // Lua: input("string")
 static int node_input( lua_State* L ) {
-  size_t l = 0;
-  const char *s = luaL_checklstring(L, 1, &l);
-  if (lua_put_line(s, l)) {
-    NODE_DBG("Result (if any):\n");
-    user_process_input(true);
-  }
+  luaL_checkstring(L, 1);
+  lua_getfield(L, LUA_REGISTRYINDEX, "stdin");
+  lua_rawgeti(L, -1, 1);             /* get the pipe_write func from stdin[1] */
+  lua_insert(L, -2);                           /* and move above the pipe ref */ 
+  lua_pushvalue(L, 1);
+  lua_call(L, 2, 0);                                     /* stdin:write(line) */
   return 0;
 }
 
-static int output_redir_ref = LUA_NOREF;
 static int serial_debug = 1;
+
 void output_redirect(const char *str) {
   lua_State *L = lua_getstate();
-  // if(c_strlen(str)>=TX_BUFF_SIZE){
-  //   NODE_ERR("output too long.\n");
-  //   return;
-  // }
-
-  if (output_redir_ref == LUA_NOREF) {
-    uart0_sendStr(str);
-    return;
-  }
+  int n = lua_gettop(L);
+  lua_pushliteral(L, "stdout");
+  lua_rawget(L, LUA_REGISTRYINDEX);                       /* fetch reg.stdout */
+  if (lua_istable(L, -1)) { /* reg.stdout is pipe */
+    if (serial_debug) {
+      uart0_sendStr(str);
+    }
+    lua_rawgeti(L, -1, 1);          /* get the pipe_write func from stdout[1] */
+    lua_insert(L, -2);                         /* and move above the pipe ref */ 
+    lua_pushstring(L, str);
+    lua_call(L, 2, 0);                               /* Reg.stdout:write(str) */
 
-  if (serial_debug != 0) {
+  } else { /* reg.stdout == nil */
     uart0_sendStr(str);
   }
-
-  lua_rawgeti(L, LUA_REGISTRYINDEX, output_redir_ref);
-  lua_pushstring(L, str);
-  lua_call(L, 1, 0);   // this call back function should never user output.
+  lua_settop(L, n);         /* Make sure all code paths leave stack unchanged */
 }
 
+extern int pipe_create(lua_State *L);
```

Comment:
I've seen a few of these now. Could we get them into a proper header instead please? :)

Reply:
Yes there are 2 explicit externs in `lua.c` and `node.c` but these are internal to VM and  I **_don't_** want to expose this extern for general application use in a general header, which is why it isn't in one.

We can do it this way or I can have a header with one line in it called, say `pipe_for_lua_runtime_only.h`.  I chose the pragmatic option.

Reply:
\<insert clever retort here\> (because I can't come up with one - it's not like that particular prototype is ever likely to change)